### PR TITLE
refactor: carry namespace name in NamespaceData

### PR DESCRIPTION
**Pull request by domodwyer**
This PR embeds the name of the namespace in the `NamespaceData`, using a ref-counted type wrapper (`NamespaceName`) for type safety (aka getting the compiler to do the work for me).

---

* refactor: carry namespace name in NamespaceData (abb9122e2)

      Changes the ingester's NamespaceData to carry a ref-counted string
      identifier as well as the ID.
      
      The backing storage for the name in NamespaceData is shared with the
      index map in ShardData, so it is effectively free!

### CODE REVIEW

File: ingester/src/data/namespace.rs [hunk 1]
```diff
@@ -432,7 +466,16 @@ mod tests {
...
+        let ns = NamespaceData::new(
+            ns_id,
+            NAMESPACE_NAME.into(),
+            shard_id,
+            partition_provider,
+            &*metrics,
+        );
+
+        // Assert the namespace name was stored
+        assert_eq!(&**ns.namespace_name(), NAMESPACE_NAME);
```
Comment (alamb) [inline ingester/src/data/namespace.rs:478]: I wonder if you would consider implementing `AsRef<str>` or something for `NamespaceName` to make this code easier on the eyes

File: ingester/src/data.rs [hunk 2]
```diff
@@ -815,7 +815,7 @@ mod tests {
 
         let (table_id, partition_id) = {
             let sd = data.shards.get(&shard1.id).unwrap();
-            let n = sd.namespace("foo").unwrap();
+            let n = sd.namespace(&"foo".into()).unwrap();
```
Comment (alamb) [inline ingester/src/data.rs:818]: I wonder if we could improve this usability a bit (and avoid the `&...into()`)

**Review by alamb — APPROVED** [review]
Looks like a good change to me.

File: ingester/src/data.rs (hunk 2 above)
Comment (domodwyer) [reply ingester/src/data.rs:818]: This will all become cleaner when it's `NamespaceName` all the way through!

Comment (alamb) [reply ingester/src/data.rs:818]: ok

**Pull request merged**